### PR TITLE
Add ability to change multiplayer queue modes

### PR DESCRIPTION
**Pull request by smoogipoo**
Preview: https://drive.google.com/file/d/1ZnrHxGMBNJu9vFtb1-A0fR2To2a0cdGY/view?usp=sharing

Three new modes arrive in multiplayer:
- Host only: The traditional multiplayer experience you're familiar with. Only the host can change the current item.
- All players: All players can enqueue items, which are played in the order they were added.
- All players (round robin): All players can enqueue items, which are played in order of fairness - the player with the least of their own items played so far gets to go first. Colloquially referred to as "host-rotate".

---

Prereqs (osu and osu-server-spectator are inter-dependent):
- [x] https://github.com/ppy/osu-web/pull/8336
- [x] https://github.com/ppy/osu-web/pull/8348
- [x] https://github.com/ppy/osu-server-spectator/pull/84

I apologise for the large PR - it's an entirely breaking change so I have to go the full way through.

Multiplayer is restructured so that:
- `MultiplayerRoomSettings` no longer contains the current map/ruleset/mods. These are now expected to be taken from the `PlaylistItem`. `MultiplayerRoomSettings` still provides the playlist item id.
- Items are added to the playlist via the new server `AddPlaylistItem()` server method.

Both of the above are breaking changes.

The `MultiplayerPlaylistItem` class has been added in order to communicate playlist items with the server. I've left in a `UserId` property, which is to be used for the upcoming fair-play mode, though initial usages will be server-side only.

`BeatmapSelectionControl` no longer exists, as it's too complicated to make it work for all usages. Instead, the playlist has been inlined:
- Into `MultiplayerMatchSubScreen`, containing the full playlist.
- Into `MultiplayerMatchSettingsOverlay`, containing only a single item, and reproducing the previous `BeatmapSelectionControl` functionality. This means that the match has to be created with a single playlist item for the time being.

The playlist has been inverted in `MultiplayerMatchSubScreen` for the time being. It's easier to handle this way without working around scroll, as the playlist length expands out of view.

Still to be done:
- [x] The host should be able to enqueue items in host-only mode.
- [x] Item owners should be able to delete their own items in free-for-all and fair-rotate modes.
- [x] Item owners should be able to edit their own items in free-for-all and fair-rotate modes.
- [x] The host should be able to delete/change items in all modes.
- [x] Split the `MultiplayerMatchSubScreen` playlist into two sections, maybe something like Apple Music (history sorted newest-first, and up-next sorted oldest-first).
- [x] Add owner to playlist item panels.
- [ ] Allow the host to change the order of items in the queue.
- [x] The displayed playlist should be ordered by the queueing mode.

As for testing - everything except fair-play mode is testable via visual tests (`TestSceneMultiplayer`) to get a feel for things.

### CODE REVIEW

File: osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs
Comment (smoogipoo) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:187]: I'll fix this in a separate PR, I want to add tests.

**Review by bdach — COMMENTED** [review]
Notes from a rudimentary read-through without running. Sorry if some of these seem obvious, I figure I'd rather ask than try to get into each detail myself given this diff size.

File: osu.Game/Online/Rooms/PlaylistItem.cs
Comment (smoogipoo) [reply osu.Game/Online/Rooms/PlaylistItem.cs:105]: I agree this is really bad. Hopefully we can improve this by splitting the api and non-api parts of the models.

I _think_ for the time being it shouldn't cause any problems since we're always calling `.Equals()` after mapping the mods across (e.g. `MultiplayerClient.createPlaylistItem()`).

File: osu.Game/Online/Multiplayer/IMultiplayerRoomServer.cs
Comment (smoogipoo) [reply osu.Game/Online/Multiplayer/IMultiplayerRoomServer.cs:83]: I don't disagree. Now that I look back at it, `MultiplayerRoom` only needs to be sent across once over the wire. Will try to make it work.

File: osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs
Comment (peppy) [reply osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs:22]: `OwnerID` sounds good (not sure the `PlaylistItem` prefix is required)

Comment (smoogipoo) [reply osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs:22]: Agree. Is `OwnerID` too short to describe it? Could add an xmldoc (was hoping to go through all multiplayer signalr classes in a followup).

**Comment by peppy** [comment]
A few initial questions, just testing high level functionality:

- Currently a user can join a room, add 1000 items and leave, making the room unplayable without a full reset. I think we discussed adding an upper limit per user - probably something we need for an initial release?
- When creating a new room, there's no active item selected (ie. no playlist item has a yellow border), is this intended (tested with free-for-all)?
- I understand the playlist list will be split in the future into history and future, but right now the active item is not tracked at all, which means it's usually off-screen. Maybe add a `ScrollIntoView` when the active item changes to bring it back onto screen?

When trying to change queue type, I got a not implemented exception:

```csharp

2021-11-16 04:34:12 [11]: Switching room ruleset to osu.Server.Spectator.Hubs.HeadToHead
fail: Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher[8]
      Failed to invoke hub method 'ChangeSettings'.
      System.NotImplementedException: The method or operation is not implemented.
         at osu.Server.Spectator.Hubs.MultiplayerQueue.updateCurrentItem(Boolean notifyHub) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerQueue.cs:line 313
         at osu.Server.Spectator.Hubs.MultiplayerQueue.ChangeMode(QueueModes newMode) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerQueue.cs:line 96
         at osu.Server.Spectator.Hubs.MultiplayerHub.ChangeSettings(MultiplayerRoomSettings settings) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 510
         at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.ExecuteMethod(ObjectMethodExecutor methodExecutor, Hub hub, Object[] arguments)
         at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.<>c__DisplayClass16_0.<<Invoke>g__ExecuteInvocation|0>d.MoveNext()
```

File: osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs
Comment (bdach) [reply osu.Game/Online/Rooms/MultiplayerPlaylistItem.cs:22]: `OwnerID` is also good with me, yep.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs:112]: It's a bit of an aside, but I've fixed in this PR because it's a one line change.

**Comment by smoogipoo** [comment]
> When creating a new room, there's no active item selected

I can't seem to reproduce this anymore, so I'm hoping some of the recent changes have resolved it.

> Maybe add a `ScrollIntoView` when the active item changes to bring it back onto screen?

I've tried, but a very simple implementation seems quite flaky, likely to do with playlist item changes requiring a remove+add (i.e. when the previous item gets expired). I'd do it in a separate PR.

> Currently a user can join a room, add 1000 items and leave, making the room unplayable without a full reset.

Sure. Any suggestions on how this should be done? Should I just hardcode a value into the server?
Also, is it something that needs to be shared with the client via some common means (e.g. a class to store the queue mode + limit)? If so that changes the implementation quite a bit.

**Comment by peppy** [comment]
I'd hardcode the limit server-side for now. 1 or 3 sounds fine. The assumption is it will be replaced with a better solution in the coming weeks, right?

**Comment by smoogipoo** [comment]
> When trying to change queue type, I got a not implemented exception:

Please retry. I couldn't repro but I would assume it's somehow a mismatch between client-server, as I used to have a fair-play branch which was required to get it working that I forgot to mention (though logs would indicate it shouldn't have had an effect). I've merged things into a single server-side branch

> The assumption is it will be replaced with a better solution in the coming weeks, right?

I still need to know how you envision this actually working - is it a setting which is passed from server-to-client, is it a server-side error that only occurs on add item, is it something that should be adjustable by the host within reason, etc.

Depending on the answer, the `MultiplayerRoomSettings` implementation has to change to contain more than just a `QueueMode` enum value.

**Comment by smoogipoo** [comment]
For the time being, I've added a per-user limit of 3 (https://github.com/ppy/osu-server-spectator/pull/84/commits/0bc500599091cdb23d7bf2345d41c37c4fa1963f).

File: osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
Comment (smoogipoo) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:122]: It actually matches the server better now. Compare this implementation to https://github.com/ppy/osu-server-spectator/blob/0bc500599091cdb23d7bf2345d41c37c4fa1963f/osu.Server.Spectator/Hubs/MultiplayerHub.cs#L587-L631, which is called after the `UserStateChanged` callback on user state changes https://github.com/ppy/osu-server-spectator/blob/0bc500599091cdb23d7bf2345d41c37c4fa1963f/osu.Server.Spectator/Hubs/MultiplayerHub.cs#L304-L326. (Yes, the `TestMultiplayerClient` doesn't have that `LoadRequested` part, it was missed but is a separate issue altogether).

The technical reason for this change is that state changes directly to `Idle` are also valid. An alternative to these changes is to add the `Idle` transition falling through to the `FinishedPlay` case, however that in combination with this code not really matching the server in the first place and having `TestSceneMultiplayerPlayer` relying on this broken behaviour to start the match once a user enters the loaded state (bypassing `StartMatch()` altogether), made me feel a bit uneasy.

Would it be better if I split it out into a separate method like the server does?

**Comment by peppy** [comment]
Testing at a functional level only, a few more things came up:

- When switching to host-only mode, it would be best if the existing queue was not cleared. Even though the host doesn't have the ability to change the playlist, let's make that the initial behaviour.
- The naming is a bit hard to follow for the mode selection. I think we could probably do with having a custom control for this (think three large buttons with an icon, main and description text explaining what the queue mode entails) but until then, how about the following:

## Queue mode

- Host only
- All players
- All players (round robin)

- The displayed playlist in the client should actually be in the order it is going to be played. Right now in round robin mode the next item will jump around depending on the order in which users add the beatmaps. This probably needs to be added to the "still to be done" section of this PR's description.

**Comment by peppy** [comment]
I'm not yet sure of the repro steps (will try and figure it out), but managed to get things in a very weird state:

```csharp

fail: Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher[8]
      Failed to invoke hub method 'AddPlaylistItem'.
      osu.Game.Online.Multiplayer.InvalidStateException: Can't enqueue more than 3 items at once.
         at osu.Server.Spectator.Hubs.MultiplayerQueue.AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerQueue.cs:line 164
         at osu.Server.Spectator.Hubs.MultiplayerHub.AddPlaylistItem(MultiplayerPlaylistItem item) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 442
         at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.ExecuteMethod(ObjectMethodExecutor methodExecutor, Hub hub, Object[] arguments)
         at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.<>c__DisplayClass16_0.<<Invoke>g__ExecuteInvocation|0>d.MoveNext()
2021-11-18 05:41:33 [11]: Switching queue mode to osu.Server.Spectator.Hubs.MultiplayerQueue
2021-11-18 05:42:21 [11]: Switching queue mode to osu.Server.Spectator.Hubs.MultiplayerQueue
2021-11-18 05:42:25 [11]: Requesting to leave room
2021-11-18 05:42:25 [11]: Leaving room 3
2021-11-18 05:42:28 [11]: Disconnected
2021-11-18 05:42:28 [11]: Disconnected
2021-11-18 05:42:35 [12]: Switching room ruleset to osu.Server.Spectator.Hubs.HeadToHead
2021-11-18 05:42:42 [12]: Switching queue mode to osu.Server.Spectator.Hubs.MultiplayerQueue
fail: Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher[8]
      Failed to invoke hub method 'ChangeSettings'.
      System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
         at System.Collections.Generic.List`1.get_Item(Int32 index)
         at osu.Server.Spectator.Hubs.MultiplayerQueue.UpdateFromQueueModeChange() in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerQueue.cs:line 79
         at osu.Server.Spectator.Hubs.MultiplayerHub.ChangeSettings(MultiplayerRoomSettings settings) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 491
         at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.ExecuteMethod(ObjectMethodExecutor methodExecutor, Hub hub, Object[] arguments)
         at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.<>c__DisplayClass16_0.<<Invoke>g__ExecuteInvocation|0>d.MoveNext()
```

Here's a clip of what I was doing at the time:

https://user-images.githubusercontent.com/191335/142359650-40a9c911-1aef-49df-a4d3-0731c147dd20.mp4

Note the first time I add a new beatmap after switching back to free add mode, it seems to get added below the current item or something similar.

**Comment by peppy** [comment]
Also with the renaming of queue modes, we should probably decide on how the classes are named and update them to be closer to expectations. Let's decide on that before actually making changes across three PRs.

**Comment by smoogipoo** [comment]
I've applied server side changes regarding the host-only mode as mentioned above. I'm not sure what your error is all about but the changes I've had to make as a result touched the area where the error seemingly occurred.

Should be impossible to have that happen now since items aren't ever cleared.

> Note the first time I add a new beatmap after switching back to free add mode, it seems to get added below the current item or something similar.

You switched to fair play mode, so what you're seeing there could be intentional... Hard to say.

**Comment by peppy** [comment]
> You switched to fair play mode, so what you're seeing there could be intentional... Hard to say.

Definitely need the user display and correct client-side queue ordering (based on play order) as priorities.

File: osu.Game/Screens/OnlinePlay/OnlinePlayComposite.cs
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/OnlinePlayComposite.cs:69]: Not a big deal, but is the `nameof` portion here required for some reason? Doesn't seem like it should be.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs:248]: I guess this is a simplification of code? Is the plan to make this further editable on this screen though?

Weird side effect which is funny (doesn't require fixing):

https://user-images.githubusercontent.com/191335/142571413-df71a675-d4e8-453d-95dc-f8bb454f0070.mp4

File: osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs:204]:
```suggestion
                                                                new Section("Queue mode")
```

As discussed

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs
Comment (peppy) [inline osu.Game/Online/Multiplayer/MultiplayerClient.cs:142]: Not too confident on this one. If a user joins a room with a large history is this going to potentially take seconds or minutes? Maybe okay for an initial release, but probably not as a permanent solution. The population of everything that isn't the beatmap id should likely happen separate from the join room operation.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
I think I'm pretty happy to go forward with how things are for these PRs. Still hoping to at very least get the user display up for the initial release but apart from the minor issues pointed out (and the remaining renames) this PR (and the related ones) are feeling haphazardly solid.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerMatchSettingsOverlay.cs:248]: The logic just doesn't work well otherwise (using `BeatmapSelectionControl`), it was much too complicated which seemed super unnecessary.

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs
Comment (smoogipoo) [reply osu.Game/Online/Multiplayer/MultiplayerClient.cs:142]: Sure, but that's more involved to do. Too out of scope for this PR, imo.

File: osu.Game.Tests/Visual/Multiplayer/QueueingModes/QueueModeTestScene.cs
Comment (smoogipoo) [reply osu.Game.Tests/Visual/Multiplayer/QueueingModes/QueueModeTestScene.cs]: I've removed the namespace for now.

File: osu.Game/Screens/OnlinePlay/OnlinePlayComposite.cs
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/OnlinePlayComposite.cs]: Not required, only required when the name of the property doesn't match the resolved name. It's a pretty hidden implementation detail...

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs
Comment (peppy) [reply osu.Game/Online/Multiplayer/MultiplayerClient.cs:142]: of course

**Comment by smoogipoo** [comment]
Renames have been applied.

**Comment by peppy** [comment]
Tests no look good.

File: osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs
Comment (bdach) [reply osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs:121]: Never mind, I missed that the switch is over *room* states, not *user* states. Subtle but important distinction, which makes this 100% okay with me.

**Comment by bdach** [comment]
I've skimmed this again (without testing) and I'm all good with the structure, so on that part I see no reason not to go forward. I may test later, but even if I find issues, I'm likely going to just open issues about it or fix myself post-merge because this diff is too large to hold up over bugs.

As for the test failures, I have a few commits that seem to fix them [on this branch](https://github.com/ppy/osu/compare/master...bdach:multi-queueing-modes-test-fixes?expand=1) ([first](https://github.com/ppy/osu/commit/9c30fc9f9b0b85c74390f898b7f6ad82d3ecc328), [second](https://github.com/ppy/osu/commit/4453479f805986aa5f8be1e5a784acb90892de5d), [third](https://github.com/ppy/osu/commit/f7829cc85e1227888fb0f5d6a4d75dd6211e6e0a) - reasonings for each change are in the extended commit message). They seem fairly innocuous but I'm not 100% about them, so I didn't push them in outright; can do so if they are considered acceptable.

**Comment by bdach** [comment]
Actually, on comparing this pull and the server-side pull, I believe there may be a discrepancy between the test multiplayer client implementation and the real one as to [how playlist items are treated when the queue mode changes](https://github.com/ppy/osu-server-spectator/pull/84#discussion_r753826570), so that may need to be addressed before this can be merged as it is potentially a pretty fundamental change in UX.

**Comment by smoogipoo** [comment]
Tests should be resolved now.

---

> I believe there may be a discrepancy between the test multiplayer client implementation and the real one as to how playlist items are treated when the queue mode changes

Indeed, fixed via https://github.com/ppy/osu/pull/15640/commits/642097166033bbcce457b09668e5c6fcdd51d5ce.

> https://github.com/ppy/osu/commit/9c30fc9f9b0b85c74390f898b7f6ad82d3ecc328

In general, the reason this was done is to not rely on the schedule having run for the server-side playlist to be up-to-date. I believe your third change (https://github.com/ppy/osu/commit/f7829cc85e1227888fb0f5d6a4d75dd6211e6e0a) may prove problematic if you add a `Task.Delay()` after the await in `PlaylistItemAdded()` (it's already on a different thread by that point).

In my changes, I've instead stored a "server-side" playlist. https://github.com/ppy/osu/pull/15640/commits/10dc08a8553be586d6b1946f3900723918d3f4fa

> https://github.com/ppy/osu/commit/4453479f805986aa5f8be1e5a784acb90892de5d

I think it's better to compare via IDs in these cases, as I've done in https://github.com/ppy/osu/pull/15640/commits/53dbbd6d64e02f84597816347f9fb97acaf9fd82.

**Comment by peppy** [comment]
I've reviewed this about as much as I can given its size. Let's get this merged and push forward. I'll do a (nuget-only) release post-merge to get the server side up-to-date and deployed on the dev server.

**Review by peppy — APPROVED** [review]

**Review by bdach — APPROVED** [review]

**Pull request merged**
